drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -2134,6 +2677,71 @@ void RasterizerCanvasGLES2::initialize() {
 		data.polygon_index_buffer_size = index_size;
 	}
 
+	// batch buffer
+	{
+		// the maximum num quads in a batch is limited by GLES2. We can have only 16 bit indices,
+		// which means we can address a vertex buffer of max size 65535. 4 vertices are needed per quad.
+
+		// Note this determines the memory use by the vertex buffer vector. max quads (65536/4)-1
+		// but can be reduced to save memory if really required (will result in more batches though)
+		int max_quads = (65536 / 4) - 1;
+
+		uint32_t sizeof_batch_vert = sizeof(BatchVert);
+
+		bdata.max_quads = max_quads;
+
+		// 4 verts per quad
+		bdata.vertex_buffer_size_units = max_quads * 4;
+
+		// the index buffer can be longer than 65535, only the indices need to be within this range
+		bdata.index_buffer_size_units = max_quads * 6;
+
+		// this comes out at approx 64K for non-colored vertex buffer, and 128K for colored vertex buffer
+		bdata.vertex_buffer_size_bytes = bdata.vertex_buffer_size_units * sizeof_batch_vert;
+		bdata.index_buffer_size_bytes = bdata.index_buffer_size_units * 2; // 16 bit inds
+
+		// create equal number of norma and colored verts (as the normal may need to be translated to colored)
+		bdata.verts.create(bdata.vertex_buffer_size_units); // 512k
+		bdata.verts_colored.create(bdata.verts.max_size()); // 1024k
+
+		// num batches will be auto increased dynamically if required
+		bdata.batches.create(1024);
+		bdata.batches_temp.create(bdata.batches.max_size());
+
+		// batch textures can also be increased dynamically
+		bdata.batch_texs.create(32);
+
+		// just reserve some space (may not be needed as we are orphaning, but hey ho)
+		glGenBuffers(1, &bdata.GL_vertex_buffer);
+		glBindBuffer(GL_ARRAY_BUFFER, bdata.GL_vertex_buffer);
+		glBufferData(GL_ARRAY_BUFFER, bdata.vertex_buffer_size_bytes, NULL, GL_DYNAMIC_DRAW);
+		glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+		// pre fill index buffer, the indices never need to change so can be static
+		glGenBuffers(1, &bdata.GL_index_buffer);
+		glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, bdata.GL_index_buffer);
+
+		Vector<uint16_t> indices;
+		indices.resize(bdata.index_buffer_size_units);
+
+		for (int q = 0; q < max_quads; q++) {
+			int i_pos = q * 6; //  6 inds per quad
+			int q_pos = q * 4; // 4 verts per quad
+			indices.set(i_pos, q_pos);
+			indices.set(i_pos + 1, q_pos + 1);
+			indices.set(i_pos + 2, q_pos + 2);
+			indices.set(i_pos + 3, q_pos);
+			indices.set(i_pos + 4, q_pos + 2);
+			indices.set(i_pos + 5, q_pos + 3);
+
+			// we can only use 16 bit indices in GLES2!
+			CRASH_COND((q_pos + 3) > 65535);
```

Comment:
Change to ``ERR_FAIL_COND_MSG``

Reply:
I can make it debug only, but this is to quickly catch logic errors if this allocation code is changed in future (i.e. if someone later changes max_quads without realising they can overrun the max index size).

If you overrun the max index size, it will 'appear' to work, however the whole vertex buffer won't get used in drawing and it will wrap around to the front causing hard to diagnose visual bugs. Best case. So really is a good idea to catch it early.